tests/test_packageAlerts.py

```diff
@@ -0,0 +1,389 @@
+
+        cutoutBytes = packageAlerts.makeCutoutBytes(cutout)
+        tempMemFile = afwFits.MemFileManager(len(cutoutBytes))
+        tempMemFile.setData(cutoutBytes, len(cutoutBytes))
+        cutoutFromBytes = afwImage.ExposureF(tempMemFile)
+        self.assertTrue(cutout, cutoutFromBytes)
+
+    def testMakeJsonAlert(self):
+        """Test stripping data from the various data products and into a
+        dictionary "alert".
+        """
+        packageAlerts = PackageAlertsTask()
+        alertId = 1234
+
+        for srcIdx, diaSource in self.diaSources.iterrows():
+            sphPoint = geom.SpherePoint(diaSource["ra"],
+                                        diaSource["decl"],
+                                        geom.degrees)
+            cutout = self.exposure.getCutout(sphPoint,
+                                             packageAlerts.cutoutBBox)
+            cutputBytes = packageAlerts.makeCutoutBytes(cutout)
+            objSources = self.diaSourceHistory.loc[srcIdx[0]]
+            alert = packageAlerts.makeJsonAlert(
+                alertId,
+                diaSource,
+                self.diaObjects.loc[srcIdx[0]],
+                objSources,
+                cutout,
+                None)
+            self.assertEqual(len(alert), 9)
+
+            self.assertEqual(alert["alertId"], alertId)
+            self.assertEqual(alert["diaSource"], diaSource.to_dict())
+            self.assertEqual(alert["cutoutDifference"]["stampData"],
+                             cutputBytes)
+            self.assertEqual(alert["cutoutTemplate"],
+                             None)
+
+    def testRun(self):
+        """Test the run method of package alerts.
+        """
+        packConfig = PackageAlertsConfig()
+        tempdir = tempfile.mkdtemp(prefix='alerts')
+        packConfig.alertWriteLocation = tempdir
+        packageAlerts = PackageAlertsTask(config=packConfig)
+
+        packageAlerts.run(self.diaSources,
+                          self.diaObjects,
+                          self.diaSourceHistory,
+                          self.exposure,
+                          None,
+                          None)
+
+        ccdVisitId = self.exposure.getInfo().getVisitInfo().getExposureId()
+        with open(os.path.join(tempdir, f"{ccdVisitId}.avro"), 'rb') as f:
+            writer_schema, data = \
+                packageAlerts.alertSchema.retrieve_alerts(f)
+        self.assertEqual(len(data), len(self.diaSources))
+        for idx, alert in enumerate(data):
+            for key, value in alert["diaSource"].items():
+                if isinstance(value, float):
+                    if np.isnan(self.diaSources.iloc[idx][key]):
+                        self.assertTrue(np.isnan(value))
+                    else:
+                        self.assertAlmostEqual(
+                            1 - value / self.diaSources.iloc[idx][key],
+                            0.)
```

Comment:
Why not `self.assertAlmostEqual(value, self.diaSources.iloc[idx][key])`?

Reply:
The code wasn't passing for some round tripped float values in an absolute sense, but does pass for relative values hence the comparison here.

Reply:
Okay, I think I understand why. `assertAlmostEqual` checks if the values are equal to 7 decimal places. Doing "almost-equal-ness" based on decimal points doesn't make much sense for floats though; for a big exponent the mantissa won't express much precision down there. This test helps by changing the calculation - it becomes "self.assertEqualToWithinOnePartPer10Million", basically, which is better.